docs/http/http-spans.md

```diff
@@ -117,7 +117,7 @@ sections below.
 |---|---|---|---|---|
 | `error.type` | string | Describes a class of error the operation ended with. [1] | `timeout`; `name_resolution_error`; `500` | Conditionally Required: If request has ended with an error. |
 | `http.request.body.size` | int | The size of the request payload body in bytes. This is the number of bytes transferred excluding headers and is often, but not always, present as the [Content-Length](https://www.rfc-editor.org/rfc/rfc9110.html#field.content-length) header. For requests using transport encoding, this should be the compressed size. | `3495` | Recommended |
-| `http.request.header.<key>` | string[] | HTTP request headers, `<key>` being the normalized HTTP Header name (lowercase, with `-` characters replaced by `_`), the value being the header values. [2] | `http.request.header.content_type=["application/json"]`; `http.request.header.x_forwarded_for=["1.2.3.4", "1.2.3.5"]` | Opt-In |
+| `http.request.header.<key>` | string[] | HTTP request headers, `<key>` being the normalized HTTP Header name (lowercase), the value being the header values. [2] | `http.request.header.content-type=["application/json"]`; `http.request.header.x-forwarded-for=["1.2.3.4", "1.2.3.5"]` | Opt-In |
```

Comment:
I think it's worth pointing out that this would be the very first convention defined by OTel that would use a hyphen in the attribute key.

The spec defines the following SHOULD guideline:
> For each multi-word dot-delimited component of the attribute name **separate the
  words by underscores (i.e. use snake_case)**. For example `http.response.status_code`
  denotes the status code in the http namespace.

-- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/attribute-naming.md

I'm not aware of any particular reason for that beyond consistency. Making an exception here for attribute key parts derived from another source (header names in this case) seems okay to me but I wonder whether any processors or consumers downstream may inaccurately depend on the assumption they'd never receive a dash. I could only think of this being problematic if subtracting attributes with a syntax of `attribute_a-attribute_b` (with the ambiguity of `-` being a minus sign) was a thing.

The only MUST requirement I could find for attribute keys is very permissive, so I think this change should be fine:

> The attribute key MUST be a non-`null` and non-empty string.

-- https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute

Reply:
> I could only think of this being problematic if subtracting attributes with a syntax of `attribute_a-attribute_b` (with the ambiguity of `-` being a minus sign) was a thing.

@open-telemetry/collector-contrib-approvers would this cause problems for OTTL? Looking at https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md#paths and https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md#math-expressions I'm thinking it wouldn't...

Reply:
> I think it's worth pointing out that this would be the very first convention defined by OTel that would use a hyphen in the attribute key.

ya this is definitely the main risk in this change to consider

Reply:
CC @evan-bradley

Reply:
It will not affect OTTL.  Since this is an attribute key it would only ever be referenced in OTTL as a string, surrounded by `""`.  Such as `set(attributes["http.request.header.content-type"], "application/json")`.

That said, I do feel it is risky to start creating exceptions that allow `-` in the attribute key.  Exceptions breed more exceptions in my opinion.

Reply:
Regarding an exception -

IIUC - we're *not* actually defining a semconv attribute, we're defining a semconv "encoding" of HTTP headers and keeping them as-is.

I.e. the convention is `http.request.header.<key>` where `<key>` is outside our convention, effectively.


Two thoughts here:

1. I'm not a fan of upper/lower-case requirements on data, unless we know that upper-lower case is simple ASCII/English standard.  i18n upper/lower-case is NOT something we want in our specification, nor should we assume it's a trivial problem.  HTTP headers are fine, as you're limited to those characters.
2. Should we avoid *any* normalization so you can distinguish conflicting headers?  If this is to truly solve #304 and we consider that an important use case, I'd suggest we don't normalize at all, but just return whatever normalization the HTTP client / server implemetnation we've instrumented provides.

Reply:
some additional thoughts around upper/lowercase:

* if you configure your http instrumentation to capture a specific header, say `content-type`, you will probably not expect to get different attribute names depending on whether the original header name was `content-type` or `Content-Type`
* however, if you configure your http instrumentation to capture ALL headers (security concerns aside), maybe it's not so surprising for case to be preserved

in general, I think the least surprising behavior would be:
* if the underlying headers are case insensitive, e.g. [HTTP headers](https://www.w3.org/Protocols/HTTP/1.0/spec.html#Message-Headers) and [gRPC metadata](https://grpc.io/docs/guides/metadata/) we should normalize to lowercase (with no other normalization)
* if the underlying headers are case sensitive (maybe some messaging protocol headers?), then do not normalize to lowercase

Reply:
> I.e. the convention is http.request.header.<key> where <key> is outside our convention, effectively.


@jsuereth this is a very helpful clarification, thank you.

Reply:
To be clear, I'm still against this change as I think it doesn't solve the underlying issue.